### PR TITLE
feat(app-platform): Create Sentry App

### PR DESCRIPTION
**Pull request by mnoble**
Adds endpoint to create a Sentry App. The Organization that owns the App must be passed in and the User making the request must be in that Org.

### CODE REVIEW

File: src/sentry/api/base.py [hunk 1]
```diff
@@ -116,7 +116,7 @@ def load_json_body(self, request):
 
         request.json_body = None
 
-        if request.META.get('CONTENT_TYPE') != 'application/json':
+        if 'application/json' not in request.content_type:
```
Comment (markstory) [inline src/sentry/api/base.py]: Do you want `startswith()` here?

File: src/sentry/api/serializers/rest_framework/sentry_app.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+class ApiScopesField(serializers.WritableField):
+    def validate(self, data):
+        valid_scopes = ApiScopes()
+        if data is None:
+            raise ValidationError('Must provide scopes')
+
+        for scope in data:
+            if scope not in valid_scopes:
+                raise ValidationError('{} not a valid scope'.format(scope))
```
Comment (markstory) [inline src/sentry/api/serializers/rest_framework/sentry_app.py]: You could have a UnicodeError here if the scope has non-ascii bytes.

File: tests/sentry/api/endpoints/test_sentry_apps.py [hunk 3]
```diff
@@ -58,3 +66,59 @@ def test_no_access_without_internal_catchall(self):
...
+        assert response.status_code == 422, response.content
+        assert 'name' in response.data['errors']
+
+    @with_feature('organizations:internal-catchall')
+    def test_missing_scopes(self):
+        self.login_as(self.user)
+        response = self._post(scopes=None)
+
+        assert response.status_code == 422, response.content
+        assert 'scopes' in response.data['errors']
```
Comment (markstory) [inline tests/sentry/api/endpoints/test_sentry_apps.py]: Should you test an invalid scope as well?

File: src/sentry/api/base.py (hunk 1 above)
Comment (mnoble) [reply src/sentry/api/base.py]: Yea, that's more clear than this. 👍

**Review by markstory — APPROVED** [review]
Looks good to me.

File: src/sentry/api/endpoints/sentry_apps.py [hunk 4]
```diff
@@ -30,3 +33,28 @@ def get(self, request):
...
+
+    @requires_feature('organizations:internal-catchall', any_org=True)
+    def post(self, request):
+        serializer = SentryAppSerializer(data=request.json_body)
+
+        if not serializer.is_valid():
+            return Response({'errors': serializer.errors}, status=422)
+
+        sentry_app = Creator.run(
+            name=request.json_body.get('name'),
```
[the remark this answers is not in the document]
Comment (markstory) [reply src/sentry/api/endpoints/sentry_apps.py]: Does the name get used in a URL context anywhere?

[the remark this answers is not in the document]
Comment (mnoble) [reply src/sentry/api/endpoints/sentry_apps.py]: It gets converted into a slug, that may be used in URLs. https://github.com/getsentry/sentry/blob/master/src/sentry/models/sentryapp.py#L89

**Pull request merged**